routing/cross_mwm_road_graph.cpp

```diff
@@ -126,38 +126,49 @@ IRouter::ResultCode CrossMwmGraph::SetFinalNode(CrossNode const & finalNode)
   return IRouter::NoError;
 }
 
-BorderCross CrossMwmGraph::FindNextMwmNode(OutgoingCrossNode const & startNode,
+BorderCross CrossMwmGraph::ConstructBorderCross(OutgoingCrossNode const & startNode,
                                            TRoutingMappingPtr const & currentMapping) const
 {
-  ms::LatLon const & startPoint = startNode.m_point;
-
   // Check cached crosses.
-  auto const it = m_cachedNextNodes.find(startPoint);
+  auto const key = make_pair(startNode.m_nodeId, currentMapping->GetMwmId());
+  auto const it = m_cachedNextNodes.find(key);
   if (it != m_cachedNextNodes.end())
   {
-    return it->second;
+    // Collision check.
```

Comment:
No need to perform additional collision check - unordered_map does this stuff for you.